### PR TITLE
JS SDK v1 with beacon feature

### PR DESCRIPTION
**Pull request by MoumitaM**
## Description of the change

> Added browsers sendBeacon feature in our v1 JS SDK. Customers now can use a single SDK to send data either via xhr or browser's sendBeacon. Default is xhr. If the load() call has the "useBeacon" parameter set to true then only the data is sent via beacon.

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

---
This change is [Reviewable](https://reviewable.io/reviews/rudderlabs/rudder-sdk-js/400)

### CODE REVIEW

File: analytics.js [hunk 1]
```diff
@@ -60,7 +60,7 @@ function enqueue(rudderElement, type) {
   if (!this.eventRepository) {
     this.eventRepository = EventRepository;
   }
-  this.eventRepository.enqueue(rudderElement, type);
+  this.eventRepository.enqueue(rudderElement, type, this.beaconPlugin);
```
Comment (saikumarrs) [inline analytics.js:63]: ?

File: analytics.js [hunk 2]
```diff
@@ -1023,6 +1023,11 @@ class Analytics {
       }
     }
 
+    if (options && options.useBeacon === true) {
+      this.eventRepository.useBeacon = options.useBeacon;
+      this.eventRepository.initializeTransportMechanism();
+    }
```
Comment (saikumarrs) [inline analytics.js:1029]: This should be done before queuing is started (Ln 985)

File: analytics.js [hunk 3]
```diff
@@ -1023,6 +1023,11 @@ class Analytics {
       }
     }
 
+    if (options && options.useBeacon === true) {
+      this.eventRepository.useBeacon = options.useBeacon;
+      this.eventRepository.initializeTransportMechanism();
```
Comment (saikumarrs) [inline analytics.js:1028]: This should be internally handled by the event repository when it is triggered to start queuing. SDK should not be aware of these details.

File: utils/EventRepository.js [hunk 4]
```diff
@@ -47,6 +48,8 @@ class EventRepository {
     this.url = "";
     this.state = "READY";
     this.batchSize = 0;
+    this.useBeacon = false;
+    this.beaconQueue = beaconQueue;
```
Comment (saikumarrs) [inline utils/EventRepository.js:52]: For Ln 63, why is `payloadQueue` initialized by default?

**Review by saikumarrs — CHANGES_REQUESTED** [review]

File: utils/EventRepository.js [hunk 5]
```diff
@@ -227,12 +230,30 @@ class EventRepository {
...
     // modify the url for event specific endpoints
     const url = this.url.slice(-1) == "/" ? this.url.slice(0, -1) : this.url;
-    // add items to the queue
-    this.payloadQueue.addItem({
-      url: `${url}/v1/${type}`,
-      headers,
-      message,
-    });
+    if (this.useBeacon) {
+      const targetUrl = `${url}/beacon/v1/batch`;
```
Comment (saikumarrs) [inline utils/EventRepository.js:234]: Can you verify if we have to send it to the Beacon batch endpoint or the regular batch endpoint works?

File: utils/storage/beaconQueue.js [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+
+  flushQueue(headers, batch, url, writekey) {
+    headers.type = "application/json";
+    batch.map((event) => {
+      event.sentAt = new Date().toISOString();
+    });
+    const data = { batch };
+    const payload = JSON.stringify(data, this.replacer);
+    const blob = new Blob([payload], headers);
+    const isPushed = navigator.sendBeacon(`${url}?writeKey=${writekey}`, blob);
```
Comment (saikumarrs) [inline utils/storage/beaconQueue.js:87]: I believe we are sending the write key to the batch endpoint for some reason. Please find this out from the server team and the possibility to replace it with the regular endpoint.

File: utils/storage/beaconQueue.js [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+
+const defaults = {
+  queue: "queue",
+  maxPayloadSize: 64 * 1000,
+};
+
+class BeaconQueue {
+  constructor() {
+    this.storage = Store;
+    this.maxItems = 10;
```
Comment (saikumarrs) [inline utils/storage/beaconQueue.js:13]: Does this have to be a hard code value?

**Review by saikumarrs — CHANGES_REQUESTED** [review]

File: utils/EventRepository.js [hunk 8]
```diff
@@ -227,12 +230,30 @@ class EventRepository {
...
+  }
+
+  initializeTransportMechanism() {
+    const sendQueueData = this.sendQueueDataForBeacon.bind(this);
+    const url = this.url.slice(-1) == "/" ? this.url.slice(0, -1) : this.url;
+    const targetUrl = `${url}/beacon/v1/batch`;
+    this.beaconQueue.url = targetUrl;
+    this.beaconQueue.writekey = this.writeKey;
+    window.addEventListener("unload", sendQueueData);
   }
```
Comment (saikumarrs) [inline utils/EventRepository.js:257]: Why do these have to be here than in the beacon lib?

File: analytics.js (hunk 1 above)
Comment (MoumitaM) [reply analytics.js:63]: By mistake, removed it

File: analytics.js (hunk 2 above)
Comment (MoumitaM) [reply analytics.js:1029]: This is used to clear the previous queue if any and destroy the queue. The name is not appropriate.

File: utils/storage/beaconQueue.js (hunk 7 above)
Comment (MoumitaM) [reply utils/storage/beaconQueue.js:13]: You mean the maxItem?

File: analytics.js (hunk 2 above)
Comment (saikumarrs) [reply analytics.js:1029]: Why are we starting the queue in the first place when the decision to use the beacon or not is not yet decided?

File: utils/EventRepository.js (hunk 8 above)
Comment (MoumitaM) [reply utils/EventRepository.js:257]: Yes can be moved to beacon lib.

File: utils/EventRepository.js (hunk 4 above)
Comment (MoumitaM) [reply utils/EventRepository.js:52]: This is for xhr, haven’t changed anything.

File: analytics.js (hunk 2 above)
Comment (MoumitaM) [reply analytics.js:1029]: I think there is some misunderstanding. Say, first time you load the page with useBeacon: true. That time a queue will be created with the timestamp. Then if you reload the page a new queue will be created. So to send data from the previous queue and destroy it this is necessary. 
I have updated the name to "initializeBeaconQueue".

File: analytics.js (hunk 1 above)
Comment (saikumarrs) [reply analytics.js:63]: Appreciate the quick response but can you respond after you push the changes? I can re-verify and resolve it.

File: utils/EventRepository.js (hunk 4 above)
Comment (saikumarrs) [reply utils/EventRepository.js:52]: We're supposed to change the implementation to initialize either XHR or Beacon based on the load options.

File: utils/storage/beaconQueue.js (hunk 7 above)
Comment (saikumarrs) [reply utils/storage/beaconQueue.js:13]: Yes

File: utils/EventRepository.js (hunk 8 above)
Comment (MoumitaM) [reply utils/EventRepository.js]: done

File: utils/storage/beaconQueue.js (hunk 7 above)
Comment (MoumitaM) [reply utils/storage/beaconQueue.js:14]: done

File: utils/EventRepository.js (hunk 4 above)
Comment (MoumitaM) [reply utils/EventRepository.js]: done

File: analytics.js (hunk 3 above)
Comment (MoumitaM) [reply analytics.js]: done

File: analytics.js (hunk 2 above)
Comment (MoumitaM) [reply analytics.js]: done

File: utils/storage/beaconQueue.js [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+    this.writekey = "";
+    this.queueName = `${defaults.queue}.${Date.now()}`;
+  }
+
+  sendQueueDataForBeacon() {
+    this.sendDataFromQueueAndDestroyQueue();
+  }
+
+  init(url, writekey, options) {
+    if (options.maxItems) this.maxItems = options.maxItems;
```
Comment (saikumarrs) [inline utils/storage/beaconQueue.js:28]: `options.maxItems` sounds vague. Let's do `options.beaconQueue.maxItems`

File: utils/storage/beaconQueue.js [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+
+  clearTimer() {
+    if (this.timeOutActive) {
+      clearTimeout(this.flushQueueTimeOut);
+      this.timeOutActive = false;
+    }
+  }
+}
+const beaconQueue = new BeaconQueue();
+export default beaconQueue;
```
Comment (saikumarrs) [inline utils/storage/beaconQueue.js:130]: Why are we creating the instance by default? Can't we create more than one queue?

File: utils/xhrModule.js [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+    this.payloadQueue.addItem({
+      url: `${url}/v1/${type}`,
+      headers,
+      message,
+    });
+  }
+}
+
+const xhr = new XHR();
+export default xhr;
```
Comment (saikumarrs) [inline utils/xhrModule.js:110]: Same query as beacon queue. Why do we have to create all the instances by default?

File: utils/EventRepository.js [hunk 12]
```diff
@@ -226,14 +69,16 @@ class EventRepository {
...
 
     // modify the url for event specific endpoints
-    const url = this.url.slice(-1) == "/" ? this.url.slice(0, -1) : this.url;
-    // add items to the queue
-    this.payloadQueue.addItem({
-      url: `${url}/v1/${type}`,
-      headers,
-      message,
-    });
+    const url = this.url.slice(-1) === "/" ? this.url.slice(0, -1) : this.url;
```
Comment (saikumarrs) [inline utils/EventRepository.js:72]: Why does this have to be done for each queue item?

File: utils/EventRepository.js [hunk 13]
```diff
@@ -226,14 +69,16 @@ class EventRepository {
...
+    if (this.useBeacon) {
+      const targetUrl = `${url}/beacon/v1/batch`;
```
Comment (saikumarrs) [inline utils/EventRepository.js:74]: Set the target URL during queue init and get rid of the `useBeacon` variable

File: analytics.js [hunk 14]
```diff
@@ -1001,6 +990,7 @@ class Analytics {
     if (serverUrl) {
```
Comment (saikumarrs) [inline analytics.js:990]: This check is redundant. Also, take writeKey and serverUrl along with the options in the `initialize` method.

File: utils/storage/beaconQueue.js (hunk 9 above)
Comment (MoumitaM) [reply utils/storage/beaconQueue.js:30]: done

File: utils/xhrModule.js [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+
+const queueOptions = {
+  maxRetryDelay: 360000,
+  minRetryDelay: 1000,
+  backoffFactor: 2,
+  maxAttempts: 10,
+  maxItems: 100,
+};
+
+class XHR {
```
[the remark this answers is not in the document]
Comment (MoumitaM) [reply utils/xhrModule.js]: done

File: utils/EventRepository.js (hunk 12 above)
Comment (MoumitaM) [reply utils/EventRepository.js]: done

File: utils/EventRepository.js (hunk 13 above)
Comment (MoumitaM) [reply utils/EventRepository.js]: done

File: analytics.js (hunk 14 above)
Comment (MoumitaM) [reply analytics.js]: done

File: utils/xhrModule.js (hunk 11 above)
Comment (MoumitaM) [reply utils/xhrModule.js]: Please check the Ln. 39
`xhrQueue.processQueueElement`
I am not sure how to call `processQueueElement` class fn from inside another function 
tried with `this.processQueueElement`.

File: utils/storage/beaconQueue.js (hunk 10 above)
Comment (MoumitaM) [reply utils/storage/beaconQueue.js:131]: why do we need more than one queue, that will complicate things more?

Comment (MoumitaM) [reply utils/storage/beaconQueue.js:131]: Done

File: utils/xhrModule.js (hunk 11 above)
Comment (MoumitaM) [reply utils/xhrModule.js]: done, found a workaround.

File: utils/EventRepository.js [hunk 16]
```diff
@@ -41,162 +18,41 @@ class EventRepository {
...
-        );
+  initialize(writeKey, url, options) {
+    let queueOptions = {};
+    let targetUrl;
+    this.writeKey = writeKey;
+    this.url = url.slice(-1) === "/" ? url.slice(0, -1) : url;
+    if (options && options.useBeacon) {
+      if (
+        options &&
+        options.beaconQueue &&
```
Comment (saikumarrs) [inline utils/EventRepository.js:35]:
```suggestion
        options.beaconQueueOptions &&
```

File: utils/EventRepository.js [hunk 17]
```diff
@@ -225,15 +81,9 @@ class EventRepository {
...
 
-    // modify the url for event specific endpoints
-    const url = this.url.slice(-1) == "/" ? this.url.slice(0, -1) : this.url;
-    // add items to the queue
-    this.payloadQueue.addItem({
-      url: `${url}/v1/${type}`,
-      headers,
-      message,
-    });
+    this.queue.enqueue(headers, message, type);
```
Comment (saikumarrs) [inline utils/EventRepository.js:84]: I don't think the headers defined in Ln67 are applicable for Beacon.

Comment (saikumarrs) [reply utils/EventRepository.js:84]: You can move the headers preparation responsibility to the queues.

File: utils/EventRepository.js [hunk 18]
```diff
@@ -41,162 +18,41 @@ class EventRepository {
    *Creates an instance of EventRepository.
    * @memberof EventRepository
    */
-  constructor(options) {
-    this.eventsBuffer = [];
+  constructor() {
     this.writeKey = "";
     this.url = "";
```
Comment (saikumarrs) [inline utils/EventRepository.js:23]: These two needn't be defined at the class level

File: utils/EventRepository.js (hunk 17 above)
Comment (MoumitaM) [reply utils/EventRepository.js:81]: Used in BeaconQueue.js --> fn flushQueue --> const blob = new Blob([payload], headers);

Comment (saikumarrs) [reply utils/EventRepository.js:81]: Beacon queue ignores the headers I believe, which is why the write key is being passed as a query parameter. Can you please verify that it is part of the original implementation?

Comment (MoumitaM) [reply utils/EventRepository.js:81]: Yes it is part of the original implementation. But not needed. new Blob([payload], headers) - only takes the data and the type. Which in this case is application/json.

**Review by saikumarrs — APPROVED** [review]

**Pull request merged**